uint32 x
uint32 y
uint32 theta
uint32 time